`#[pymethods]` generated code triggers `unsafe_op_in_unsafe_fn` lint

### Bug Description

Functions in `impl` blocks annotated with `pymethods` trigger the `unsafe_op_in_unsafe_fn` lint. In a crate with `#![deny(unsafe_op_in_unsafe_fn)]` set, I get the following error:

```
error[E0133]: call to unsafe function `pyo3::impl_::pymethods::BoundRef::<'a, 'py, pyo3::PyAny>::ref_from_ptr` is unsafe and requires unsafe block
  --> src/lib.rs:10:19
   |
10 |     fn do_nothing(&self) {}
   |                   ^ call to unsafe function
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: consult the function's documentation for information on how to avoid undefined behavior
note: an unsafe function restricts its caller, but its body is safe by default
  --> src/lib.rs:8:1
   |
8  | #[pymethods]
   | ^^^^^^^^^^^^
note: the lint level is defined here
  --> src/lib.rs:1:9
   |
1  | #![deny(unsafe_op_in_unsafe_fn)]
   |         ^^^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0133`.
error: could not compile `repro_pyo3` (lib) due to 1 previous error
```

This points to a call to `ref_from_ptr`, which we can see in the expanded code:

```rust
#[doc(hidden)]
#[allow(non_snake_case)]
impl Foo {
    unsafe fn __pymethod_do_nothing__<'py>(
        py: ::pyo3::Python<'py>,
        _slf: *mut ::pyo3::ffi::PyObject,
    ) -> ::pyo3::PyResult<*mut ::pyo3::ffi::PyObject> {
        let _slf_ref = &_slf;
        let function = Foo::do_nothing;
        #[allow(clippy::let_unit_value)]
        let mut holder_0 = ::pyo3::impl_::extract_argument::FunctionArgumentHolder::INIT;
        let gil_refs_checker_0 = ::pyo3::impl_::deprecations::GilRefs::new();
        let result = {
            let ret = function(::pyo3::impl_::deprecations::inspect_type(
                ::pyo3::impl_::extract_argument::extract_pyclass_ref::<Foo>(
                    // vvv right here
                    ::pyo3::impl_::pymethods::BoundRef::ref_from_ptr(py, &_slf).0, // <<< right here
                    // ^^^ right here
                    &mut holder_0,
                )?,
                &gil_refs_checker_0,
            ));
            ::pyo3::impl_::wrap::map_result_into_ptr(
                py,
                ::pyo3::impl_::wrap::OkWrap::wrap(ret)
                    .map_err(::core::convert::Into::<::pyo3::PyErr>::into),
            )
        };
        gil_refs_checker_0.function_arg();
        result
    }
}
```

### Steps to Reproduce

Quick method:

1. Clone this repo: https://github.com/nsunderland1/repro_pyo3/tree/main
2. Run `cargo build`

Alternatively:

1. Create a PyO3 project following the steps in https://github.com/PyO3/pyo3?tab=readme-ov-file#using-rust-from-python 
4. Replace the contents of `lib.rs` with the following code:
```rust
#![deny(unsafe_op_in_unsafe_fn)]

use pyo3::prelude::*;

#[pyclass]
struct Foo;

#[pymethods]
impl Foo {
    fn do_nothing(&self) {}
}
```
3. Run `cargo build`

### Backtrace

_No response_

### Your operating system and version

macOS Sonoma

### Your Python version (`python --version`)

Python 3.10.14

### Your Rust version (`rustc --version`)

rustc 1.80.0 (051478957 2024-07-21)

### Your PyO3 version

0.22.0

### How did you install python? Did you use a virtualenv?

pyenv, using a virtual env as in the steps laid out here https://github.com/PyO3/pyo3?tab=readme-ov-file#using-rust-from-python

### Additional Info

_No response_

--- a/newsfragments/4674.fixed.md
+++ b/newsfragments/4674.fixed.md
@@ -1,0 +1,1 @@
+Fixes unintentional `unsafe_op_in_unsafe_fn` trigger by adjusting macro hygiene.

--- a/pyo3-macros-backend/src/method.rs
+++ b/pyo3-macros-backend/src/method.rs
@@ -270,9 +270,9 @@ impl FnType {
                     ::std::convert::Into::into(
                         #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(#py, &*(&#slf as *const _ as *const *mut _))
                             .downcast_unchecked::<#pyo3_path::types::PyType>()
-                    ),
+                    )
                 };
-                Some(ret)
+                Some(quote! { unsafe { #ret }, })
             }
             FnType::FnModule(span) => {
                 let py = syn::Ident::new("py", Span::call_site());
@@ -283,9 +283,9 @@ impl FnType {
                     ::std::convert::Into::into(
                         #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(#py, &*(&#slf as *const _ as *const *mut _))
                             .downcast_unchecked::<#pyo3_path::types::PyModule>()
-                    ),
+                    )
                 };
-                Some(ret)
+                Some(quote! { unsafe { #ret }, })
             }
             FnType::FnNew | FnType::FnStatic | FnType::ClassAttribute => None,
         }
@@ -332,6 +332,8 @@ impl SelfType {
         let py = syn::Ident::new("py", Span::call_site());
         let slf = syn::Ident::new("_slf", Span::call_site());
         let Ctx { pyo3_path, .. } = ctx;
+        let bound_ref =
+            quote! { unsafe { #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(#py, &#slf) } };
         match self {
             SelfType::Receiver { span, mutable } => {
                 let method = if *mutable {
@@ -344,7 +346,7 @@ impl SelfType {
                 error_mode.handle_error(
                     quote_spanned! { *span =>
                         #pyo3_path::impl_::extract_argument::#method::<#cls>(
-                            #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(#py, &#slf).0,
+                            #bound_ref.0,
                             &mut #holder,
                         )
                     },
@@ -355,7 +357,7 @@ impl SelfType {
                 let pyo3_path = pyo3_path.to_tokens_spanned(*span);
                 error_mode.handle_error(
                     quote_spanned! { *span =>
-                        #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(#py, &#slf).downcast::<#cls>()
+                        #bound_ref.downcast::<#cls>()
                             .map_err(::std::convert::Into::<#pyo3_path::PyErr>::into)
                             .and_then(
                                 #[allow(unknown_lints, clippy::unnecessary_fallible_conversions)]  // In case slf is Py<Self> (unknown_lints can be removed when MSRV is 1.75+)
@@ -665,14 +667,14 @@ impl<'a> FnSpec<'a> {
                     FnType::Fn(SelfType::Receiver { mutable: false, .. }) => {
                         quote! {{
                             #(let #arg_names = #args;)*
-                            let __guard = #pyo3_path::impl_::coroutine::RefGuard::<#cls>::new(&#pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &_slf))?;
+                            let __guard = unsafe { #pyo3_path::impl_::coroutine::RefGuard::<#cls>::new(&#pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &_slf))? };
                             async move { function(&__guard, #(#arg_names),*).await }
                         }}
                     }
                     FnType::Fn(SelfType::Receiver { mutable: true, .. }) => {
                         quote! {{
                             #(let #arg_names = #args;)*
-                            let mut __guard = #pyo3_path::impl_::coroutine::RefMutGuard::<#cls>::new(&#pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &_slf))?;
+                            let mut __guard = unsafe { #pyo3_path::impl_::coroutine::RefMutGuard::<#cls>::new(&#pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &_slf))? };
                             async move { function(&mut __guard, #(#arg_names),*).await }
                         }}
                     }
@@ -862,11 +864,13 @@ impl<'a> FnSpec<'a> {
                             _args: *mut #pyo3_path::ffi::PyObject,
                         ) -> *mut #pyo3_path::ffi::PyObject
                         {
-                            #pyo3_path::impl_::trampoline::noargs(
-                                _slf,
-                                _args,
-                                #wrapper
-                            )
+                            unsafe {
+                                #pyo3_path::impl_::trampoline::noargs(
+                                    _slf,
+                                    _args,
+                                    #wrapper
+                                )
+                            }
                         }
                         trampoline
                     },

--- a/pyo3-macros-backend/src/module.rs
+++ b/pyo3-macros-backend/src/module.rs
@@ -449,7 +449,7 @@ fn module_initialization(
             #[doc(hidden)]
             #[export_name = #pyinit_symbol]
             pub unsafe extern "C" fn __pyo3_init() -> *mut #pyo3_path::ffi::PyObject {
-                #pyo3_path::impl_::trampoline::module_init(|py| _PYO3_DEF.make_module(py))
+                unsafe { #pyo3_path::impl_::trampoline::module_init(|py| _PYO3_DEF.make_module(py)) }
             }
         });
     }

--- a/pyo3-macros-backend/src/params.rs
+++ b/pyo3-macros-backend/src/params.rs
@@ -79,7 +79,7 @@ pub fn impl_arg_params(
             .collect();
         return (
             quote! {
-                let _args = #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &_args);
+                let _args = unsafe { #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &_args) };
                 let _kwargs = #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr_or_opt(py, &_kwargs);
                 #from_py_with
             },
@@ -301,9 +301,10 @@ pub(crate) fn impl_regular_arg_param(
         }
     } else {
         let holder = holders.push_holder(arg.name.span());
+        let unwrap = quote! {unsafe { #pyo3_path::impl_::extract_argument::unwrap_required_argument(#arg_value) }};
         quote_arg_span! {
             #pyo3_path::impl_::extract_argument::extract_argument(
-                #pyo3_path::impl_::extract_argument::unwrap_required_argument(#arg_value),
+                #unwrap,
                 &mut #holder,
                 #name_str
             )?

--- a/pyo3-macros-backend/src/pymethod.rs
+++ b/pyo3-macros-backend/src/pymethod.rs
@@ -1182,25 +1182,26 @@ fn extract_object(
     let Ctx { pyo3_path, .. } = ctx;
     let name = arg.name().unraw().to_string();
 
-    let extract =
-        if let Some(from_py_with) = arg.from_py_with().map(|from_py_with| &from_py_with.value) {
-            quote! {
-                #pyo3_path::impl_::extract_argument::from_py_with(
-                    #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &#source_ptr).0,
-                    #name,
-                    #from_py_with as fn(_) -> _,
-                )
-            }
-        } else {
-            let holder = holders.push_holder(Span::call_site());
-            quote! {
-                #pyo3_path::impl_::extract_argument::extract_argument(
-                    #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &#source_ptr).0,
-                    &mut #holder,
-                    #name
-                )
-            }
-        };
+    let extract = if let Some(from_py_with) =
+        arg.from_py_with().map(|from_py_with| &from_py_with.value)
+    {
+        quote! {
+            #pyo3_path::impl_::extract_argument::from_py_with(
+                unsafe { #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &#source_ptr).0 },
+                #name,
+                #from_py_with as fn(_) -> _,
+            )
+        }
+    } else {
+        let holder = holders.push_holder(Span::call_site());
+        quote! {
+            #pyo3_path::impl_::extract_argument::extract_argument(
+                unsafe { #pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &#source_ptr).0 },
+                &mut #holder,
+                #name
+            )
+        }
+    };
 
     let extracted = extract_error_mode.handle_error(extract, ctx);
     quote!(#extracted)

--- a/tests/ui/forbid_unsafe.rs
+++ b/tests/ui/forbid_unsafe.rs
@@ -1,4 +1,5 @@
 #![forbid(unsafe_code)]
+#![forbid(unsafe_op_in_unsafe_fn)]
 
 use pyo3::*;
 

